### ISSUE TITLE
sql: Allow wipe of a cluster in-place

### ISSUE THREAD

**Issue by dt** [cockroachdb/cockroach#54505]
Currently we support dropping individual objects with commands like `DROP TABLE` and `DROP DATABASE`, and graphs of such objects with `DROP... CASCADE`.

However in some cases an operator wants to "start fresh" i.e. with no user-created objects, for example before performing a cluster restore. In our own testing we typically just stand-up a new cluster and destroy the old one in such cases, but for many production operators, this is not so easy: procurement of hardware / resources, licensing, or other organizational policies and procedures might make this extremely burdensome and effectively force wiping a cluster in-place instead.  Currently doing so is a manual process, wherein operators must find and drop objects individually. 

Additionally, just `drop`ing  all objects is not sufficient to "start fresh" as they must also be GC'ed after the configured TTL before they are truly gone.

To make this easier for operators, we should instead provide a "wipe cluster" command, e.g. `DROP ALL SCHEMA AND DATA`. 

Such a tool would:
 - Require no jobs be running.
 - Delete all user-defined descriptors.
-  Delete all jobs.
 - Delete all rows from the cursed system.comments table.
 - ClearRange the entire user table keyspan.
- (maybe ClearRange the user descs too, to prevent time-travel?)
- any other system tables that refer to user tables?

We likely want to leave system.users alone, so existing users can then go to work re-populating the now clean cluster.

It is probably reasonable to ask the operator to ensure this is only run absent other traffic, though we may also want to add a wait after deleting descriptors and/or step them though offline (though that could be expensive for lots of descs, and we don't care _much_ about correctness of a SQL write when we're about to ClearRange the whole cluster).

**Comment by pbardea** [comment]
I was thinking about what the behaviour should be for when the command was issued in an arbitrary user database that isn't defaultdb. I think we would need to create a defaultdb and switch to it. After considering some alternatives (e.g. CLI command instead of SQL, 2 phase command), I think this is the simplest and most straightforward. I'm curious if others have thoughts otherwise.

Cluster restore currently uses the `defaultdb` we create at startup as a type of admin database where we can run cluster restore. Specifically, it relies on `defaultdb` having a particular ID. I'm not a fan of this restriction, and I would like to enable cluster restore to run from any empty database.

As a note, I briefly considered if there would be advantages in handling this scenario if the command was issued at the CLI level, rather than through SQL but we still need to keep around a `defaultdb` for the user to connect to and run the cluster restore.

**Comment by pbardea** [comment]
Circling back to this, I think the scope has changed a bit. One thing to note is that I think this can be scripted with something like:
```
for database in [SHOW DATABASES]:
   ALTER DATABASE CONFIGURE ZONE USING gc.ttlseconds = 1;
   DROP DATABASE <database> CASCADE;
```

This won't check that there aren't any jobs running, but that shouldn't block things like a cluster restore. A script like this would likely be slower than just clear-ranging the data and manually dropping the descriptors, but it would be built upon already tested features -- and also minimize the number of ways to drop descriptors that we would need to maintain. I also don't think that we have an indication that performance is a bottleneck for users that want to wipe their cluster today.

If performance does become a bottleneck, perhaps it would be easiest to internal implement the above script and just preface it with a clear range across all user data so that no large deletion operations would be needed by the GC job.

One thing to note is that existing descriptors that have been dropped but are waiting to GC'd would not be dropped with this script, but I think that's a larger issue and better solved by tackling https://github.com/cockroachdb/cockroach/issues/47437 directly.

----


I am curious what what thoughts SQL schema has on how we would support such a command directly, and if the above script work-around would be feasible.

cc @ajwerner @lucy-zhang

**Comment by lucy-zhang** [comment]
If we just want to kick off GC for everything, shouldn't `ALTER RANGE default CONFIGURE ZONE` be sufficient, even if we still can't control individual tables/databases?

**Comment by lucy-zhang** [comment]
Never mind, `ALTER RANGE default` doesn't work if any of the previously dropped descriptors have some non-default value.

Another problem is that we'd presumably want to keep `defaultdb` around while dropping everything in it. It would also possible to either drop the current database or to set the current database to `""`, but both of these are kind of weird states.

It's not clear to me whether just dropping all the databases produces the desired state for the case where we need to do full-cluster restore. If we can't get rid of all the existing descriptors because we can't initiate GC, then we can't restore the new descriptors, since the IDs may be reused, right? It seems like there's some complexity involved in this problem depending on exactly what state we want to clear and for what purpose. Do we have a list of use cases in mind?

**Comment by fabiog1901** [comment]
A customer has expressed interest in having this RFE implemented, see [ZD 8667](https://cockroachdb.zendesk.com/agent/tickets/8667), and added the following wish: they would like to have a RESTORE option whereby, regardless of the current cluster state, CockroachDB will take whatever actions are required to perform the full cluster restore.

Something like `RESTORE ... WITH force`

**Comment by shermanCRL** [comment]
@fabiog1901 That’s a helpful, specific use case -- the RESTORE does the wipe. Not _sure_ that should there should be two operations overloaded into a single command.

Stepping back, currently the practice SREs use is simply stop the db process, delete the data directories, and start the server. This is complete, gives the operators good control and visibility, they can use their automation tools, no surprises.

By contrast, it seems like wipe-as-a-DB command would be messier. We’d need to maintain a full list of all the things that need to be deleted. Seems like it would be easy to miss something, now and in the future.

So a question I have is, how is a wipe by the database server be better than this SRE approach? I see two possible points mentioned above:

- @dt mentions that having the users in-place
- @fabiog1901 mentions that it gives operators a single command

**Comment by shermanCRL** [comment]
Another SRE-ish approach would be simply to restart the server with a different `--store` argument, no?

**Comment by postamar** [comment]
I agree with @shermanCRL and given the lack of recent activity I'm going to close this. Please feel free to reopen if anyone feels they can make a case that the benefits of adding this feature outweigh the many person-weeks this is going to take to implement properly.

For a start, we would need to specify the requirements more clearly and even that is not easy. Should we reset the descriptor ID counter? In fact, which data in the system database should be nuked? If we disable time travel how do we lock the cluster while it's undergoing these changes? What if the cluster is in a mixed-version state, will that be a problem? Should this statement cancel all the jobs? What about non-cancellable jobs? Even if these questions were easy to answer (to me they seem very much not obvious) we'd need a lot of testing to convince ourselves that we haven't forgotten to reset something important.
